scanner/metadata/cuesheet/cuesheet.go

```diff
@@ -0,0 +1,675 @@
+package cuesheet
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/navidrome/navidrome/log"
+)
+
+const (
+	delims          = "\t\n\r "
+	eol             = "\n"
+	framesPerSecond = 75
+	remGenre        = "GENRE"
+	remComment      = "COMMENT"
+	remDate         = "DATE"
+	remDiskID       = "DISCID"
+	remRGAlbumGain  = "REPLAYGAIN_ALBUM_GAIN"
+	remRGAlbumPeak  = "REPLAYGAIN_ALBUM_PEAK"
+	remRGTrackGain  = "REPLAYGAIN_TRACK_GAIN"
+	remRGTrackPeak  = "REPLAYGAIN_TRACK_PEAK"
+	remDiscNumber   = "DISCNUMBER"
+	remTotalDiscs   = "TOTALDISCS"
+)
+
+// Frame represent track frame
+type Frame uint64
+
+// Flags for track
+type Flags int
+
+// RemData data from REM lines
+type RemData map[string]string
+
+const (
+	// None - no flags
+	None   Flags = iota
+	Dcp          = 1 << iota
+	FourCh       = 1 << iota
+	Pre          = 1 << iota
+	Scms         = 1 << iota
+)
+
+var (
+	ErrorFrameFormat = fmt.Errorf("invalid frame format")
+)
+
+// TrackIndex data
+type TrackIndex struct {
+	Number uint
+	Frame  Frame
+}
+
+func frameFromString(str string) (Frame, error) {
+	v := strings.Split(str, ":")
+	if len(str) < 8 {
+		return 0, ErrorFrameFormat
+	}
+	if len(v) == 3 {
+		mm, _ := strconv.ParseUint(v[0], 10, 32)
+		ss, _ := strconv.ParseUint(v[1], 10, 32)
+		if ss > 59 {
+			return 0, ErrorFrameFormat
+		}
+		ff, _ := strconv.ParseUint(v[2], 10, 32)
+		if ff >= framesPerSecond {
+			return 0, ErrorFrameFormat
+		}
+		return Frame((mm*60+ss)*framesPerSecond + ff), nil
+	}
+	return 0, ErrorFrameFormat
+}
+
+func (frame Frame) Duration() time.Duration {
+	seconds := time.Second * time.Duration(frame/framesPerSecond)
+	milliseconds := time.Millisecond * time.Duration(float64(frame%framesPerSecond)/0.075)
+	return seconds + milliseconds
+}
+
+func (frame Frame) String() string {
+	seconds := frame / framesPerSecond
+	minutes := seconds / 60
+	seconds %= 60
+	ff := frame % framesPerSecond
+	return fmt.Sprintf("%.2d:%.2d:%.2d", minutes, seconds, ff)
+}
+
+// Track instance
+type Track struct {
+	Rem           RemData
+	TrackNumber   uint
+	TrackDataType string
+	Flags         Flags
+	ISRC          string
+	Title         string
+	Performer     string
+	SongWriter    string
+	PreGap        Frame
+	PostGap       Frame
+	Index         []TrackIndex
+}
+
+// File instance
+type File struct {
+	FileName string
+	FileType string
+	Tracks   []Track
+}
+
+// Cuesheet instance
+type Cuesheet struct {
+	Rem        RemData
+	Catalog    string
+	CdTextFile string
+	Title      string
+	Performer  string
+	SongWriter string
+	Pregap     Frame
+	Postgap    Frame
+	File       []File
+}
+
+func (track *Track) GetStartOffset() time.Duration {
+	if len(track.Index) > 1 {
+		return track.Index[1].Frame.Duration()
+	}
+	return track.Index[0].Frame.Duration()
+}
+
+// ReadFromFile load CUESheet from file
+func ReadFromFile(filePath string) (*Cuesheet, error) {
```

Comment:
Have you tested this well with malformed cue sheets (and even gibberish/binary files) to make sure it will return sensible errors and not crash Navidrome?

Reply:
I made some tests, but sure,need to add more cases.